New tutorial: Activating DNSSEC for Cloud DNS domains

dupuy opened the pull request:
The (currently alpha) [DNSSEC feature for Google Cloud DNS](https://groups.google.com/forum/#!msg/cloud-dns-announce/c7MD0CslHaM/5xf6RXLdCQAJ) lets users enable DNSSEC signing for their managed zones, but until they get a DS record for their domain added to the parent TLD, DNSSEC signatures aren't actually used.

The process for adding a DS record is extremely technical and different for every domain registrar that supports DNSSEC (many do not). This tutorial walks you through preliminary steps to make sure it is safe to add a DS record, and then provides links to third-party documentation that explains how to do this for many of the most popular domain registrars.

This was originally part of the Cloud DNS documentation for DNSSEC, but the Cloud docs team (rightly) pointed out that most of the content documented processes for other companies' services (the domain registrars) and that it really belonged in a community site where those others (or their customers) could provide updates and corrections. The Tutorials feature of the Google Cloud Community site was chosen as the best place to host this.

There's a lot of content in this, but it has already gone through several rounds of review by the Cloud docs team in its former incarnation as part of the Google Cloud DNS DNSSEC documentation. I have updated and checked status for all the domain registrars, validated links (the only broken one I found was actually for Google Domains) and added a few more domain registrars as well. I'm happy to incorporate any feedback (and I am pointing a few developers and tech writers inside Google at this PR for comments as well), but would like to get this up so that it isn't blocking the launch of the Google Cloud DNS DNSSEC documentation.

You can reach me at alexdupuy@google.com by e-mail or Hangouts, if you have questions about any of this.

1. [ ] Tech writing / styleguide review - @aikidave 
2. [ ] Subject matter review - @cdeccio
3. [ ] Formatting review for publishing - @jmdobry

jmdobry commented:
I'll wait for @aikidave's review before doing my own.

aikidave commented:
Will review today.

On Oct 10, 2017 3:08 PM, "Jason Dobry" <notifications@github.com> wrote:

> I'll wait for @aikidave <https://github.com/aikidave>'s review before
> doing my own.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/community/pull/260#issuecomment-335622923>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Acwz4x9KXK_TCsJZ0NKQxgN4Vdu4VITQks5sq-rggaJpZM4Pzjbn>
> .
>

dupuy commented:
If either @danigrant, who wrote much of the similar documentation for [CloudFlare](https://support.cloudflare.com/hc/en-us/articles/209114378), or @cdeccio, author of DNSViz ([dnsviz/dnsviz](https://github.com/dnsviz/dnsviz) – a tool for diagnosing DNSSEC) would be willing to review this tutorial, they are both well qualified as subject matter experts.

If neither of them is available, other potential subject matter expert reviewers would include @vdukhovni or @miekg.

dupuy commented:
Since no review yet, I rebased this onto master and fixed a problem with table formatting introduced by running mdformat on the source.

I'm assuming that if some table syntax is not recognized by GitHub native preview mode, it will have the same problem in the Google Cloud Community website, but even if the Markdown parser used by the Community hosting allows the new ':' multi-line cell syntax, the version that keeps the cells on a single line looks fine, so it's a better choice either way.

jmdobry commented on tutorials/dnssec-cloud-dns-domains/index.md:
```diff
@@ -0,0 +1,1074 @@
+and some two-letter country code (ccTLD) top-level domains, the TLD registry may
+also provide domain registrar services.
+
+<dl>
```
Any and all HTML-like content in the markdown files are stripped out when they are published, you'll have to remove the HTML in your tutorial.

dupuy replied:
Done. As this is not mentioned anywhere in the [Style Guide](https://cloud.google.com/community/tutorials/styleguide#general-style-guidelines) should I submit a separate PR to add a "Markdown usage guidelines" section that notes this restriction, as well as the specific Markdown flavor (GFM)? Other things could be added to this section later, but these are likely to be issues for other submitters as well.

jmdobry replied:
Sure, we support plain markdown, no special flavor.

One thing of note: code fences don't work within lists.

jmdobry commented on tutorials/dnssec-cloud-dns-domains/index.md:
```diff
@@ -0,0 +1,1074 @@
+If you cannot reach the ICANN TLD DNSSEC site for any reason, you can check with
+the [Verisign DNSSEC debugger][8]: just enter the TLD name and hit return; if
+you see any red ⊗ errors the TLD does not support DNSSEC (e.g.
+<http://dnssec-debugger.verisignlabs.com/aq>).
```
This would be stripped out, it looks like HTML

dupuy replied:
Fixed.

jmdobry commented on tutorials/dnssec-cloud-dns-domains/index.md:
```diff
@@ -0,0 +1,1074 @@
+This shell function checks that all authoritative servers for a domain are
+serving DNSSEC data:
+
+[embedmd]:# (checksigned.sh /checksigned/ $)
```
You'll need to run `embedmd` yourself, or I can do it after all reviews are complete.

dupuy replied:
This could also be noted in the [Style Guide](https://cloud.google.com/community/tutorials/styleguide#option-2), I can add that to the other PR mentioned above as well.

jmdobry replied:
Go ahead

jmdobry commented on tutorials/dnssec-cloud-dns-domains/index.md:
```diff
@@ -0,0 +1,1074 @@
+
+[embedmd]:# (checksigned.sh /checksigned/ $)
+
+~~~console
```
s/~~~/```/

apply throughout document

dupuy replied:
Done.

jmdobry commented:
@dupuy Looks like there are still some outstanding comments to be addressed?

dupuy commented:
To address points 1 and 3 in @aikidave's comments has required some rather drastic rewriting (I have found [Hemingway](http://www.hemingwayapp.com/) useful for this). So it's taking a while. I have pushed a commit for the first section of this rewriting, but it will take longer to complete this for the whole tutorial. Any feedback on what I've already done is welcome, as well as general comments on the style changes in the rewrite.

dupuy commented on tutorials/dnssec-cloud-dns-domains/index.md:
```diff
@@ -0,0 +1,1074 @@
+non-existent domains without additional queries to your name servers, which
+might reduce the total number of queries (especially for denial-of-service
+attacks that send millions of queries for random non-existent subdomains).
+
```
As I understand it, the purpose of the "Cost" section in these tutorials is not to do an exhaustive analysis of the factors that might influence the cost of operating a large cloud deployment, but rather to give the reader some idea of what it might actually cost them to try out the tutorial's examples.

It is probably already going a bit far to even mention the possibility of some potential benefit if you use NSEC denial of existence (the default for Google Cloud DNS when enabling DNSSEC is NSEC3). I think the simpler (and I hope now less awkward) description is sufficient for this tutorial.

I might try to write up another tutorial explaining in detail how to set up a multi-level zone structure with multiple DNSSEC signed zones: one using NSEC at the top level (containing all the names you don't care if anyone knows, like mail, www, etcetera) and others using NSEC3 below that where you can host trial features and other "unpublished" names where the fig leaf of NSEC3 might justify its use. In that tutorial, it would be appropriate to provide a detailed analysis such as you wrote (although probably not in the "Cost" section), so I will keep your suggestions in mind for that.

dupuy commented on tutorials/dnssec-cloud-dns-domains/index.md:
```diff
@@ -0,0 +1,1162 @@
+If you cannot reach the ICANN TLD DNSSEC site for any reason, you can check with
+the [Verisign DNSSEC debugger][8]: just enter the TLD name and hit return; if
+you see any red ⊗ errors the TLD does not support DNSSEC (e.g.
+http://dnssec-debugger.verisignlabs.com/aq).
```
I would argue that a TLD does not truly support DNSSEC if it has any DNSSEC validation errors, and offer the .MM (Myanmar, formerly Burma) and (more recently) .MG (Madagascar) ccTLDs as prime examples of TLDs with chronic mismanagement of DNSSEC (as documented at https://ianix.com/pub/dnssec-outages.html). There is little point in activating DNSSEC for domains in those TLDs since even though they nominally support DNSSEC, since your domain will only be accessible from non-validating resolvers for anywhere from days to weeks at a time. 

I love (really ❤️💛💚💙💜!) the DNSViz graphical display for diagnosing subtle problems, but for the purposes of this tutorial it is much easier to explain to domain administrators with limited understanding of DNSSEC how to interpret the default display of the Verisign DNSSEC debugger page ("if you see any red ⊗"). I still remember how fascinating but daunting it was to look at DNSViz when I first enabled DNSSEC for my dupuy.us domain—it simply provided more information than I was able to comprehend.

cdeccio replied:
Fair enough :)

Since you're already having them use shell scripts and the like, you could still have them try something like: `dig -t soa +ad mg @8.8.8.8 | grep -q '^;; flags:[^;]\+ad' && echo 'DNSSEC available' || echo 'DNSSEC unavailable'` as a possible alternative or in addition to the Web test.  It's not critical, I'm just thinking of ways to boil it down.

dupuy replied:
I thought about providing another shell script for this, but as it is already merely an alternative to looking up the TLD status on the [ICANN page](http://stats.research.icann.org/dns/tld_report/), it seemed overkill to provide a third script.

I already feel like the two existing shell script checks are an unfortunate bar to Windows-only users, but I don't see any better alternatives for those two cases. For this one the web-based alternatives seem preferable and more accessible.

dupuy commented on tutorials/dnssec-cloud-dns-domains/index.md:
```diff
@@ -0,0 +1,1162 @@
+resolvers to see those updates. If you activate DNSSEC at your domain registrar
+before that is complete, validating resolvers may fail to resolve names for your
+domain until it is fully propagated.
+
```
These are very good suggestions, do you think it would be better to do this with more comments in the embedded code or in the explanatory text in the tutorial? I'm inclined to the latter (explanatory text) as there are domain administrators who only use Windows systems.

Google Cloud does provide a free Linux shell environment ([Google Cloud Shell](https://cloud.google.com/shell/)) where you can run these commands (and I will add text that explicitly calls that out) but a Windows-only domain administrator is less likely to read the script than to just run it.

As for checking parent/child NS record consistency, the script uses only the NS records provided by the TLD parent; the exhortation is not to check for parent/child consistency, but to make sure that the Google Cloud DNS name servers are the (only) authoritative servers (since no others will possess the same private DNSSEC keys). I'll see if fixing the script to check NS records returned by the NS listed in the referral response is feasible.

cdeccio replied:
To be honest, I would explain both in the tutorial as well as in the code.  I don't think there's that much more explaining to do (maybe a few sentences), but sufficient background can help prevent a user from shooting themselves in the foot by skipping a few steps because they think they got it.

dupuy commented on tutorials/dnssec-cloud-dns-domains/index.md:
```diff
@@ -0,0 +1,1162 @@
+
+[11]: https://www.nameisp.com/
+[12]: https://www.name.com
+
```
There is no point in creating DS records unless the domain is already properly delegated with NS records. It is listed as a pre-requisite (Before you begin): "have access to the delegated Google Cloud DNS managed zone serving that domain" but I will make it more explicit and see if there is a checker that provides a clear display of authoritative servers.

Maihistory replied:
> Do you have any way of doing pre-delegation testing?  I know the managed DNS is under Google's control, so maybe it's moot.  But there are tools for doing this, e.g., https://github.com/dnsviz/dnsviz#pre-deployment-dns-testing

dupuy commented on tutorials/dnssec-cloud-dns-domains/index.md:
```diff
@@ -0,0 +1,1162 @@
+debugger][14]: just enter your domain name and hit return; you should only see
+green checks (e.g. http://dnssec-debugger.verisignlabs.com/publicdns.goog).
+
+[14]: http://dnssec-debugger.verisignlabs.com/us
```
Shameless plug is still a point well taken. I will add a link to DNSViz and tell them to look for the blue arrow (and no red boxes on the left).

cdeccio replied:
thanks :)

aikidave commented:
Is this ready for me to have another look? Happy to help however I can!

jmdobry commented:
@dupuy What remains to be addressed in this tutorial before it can be published?

jmdobry commented:
Published: https://cloud.google.com/community/tutorials/dnssec-cloud-dns-domains
